common/persistence/elasticsearch/visibility_store.go

```diff
@@ -925,30 +882,158 @@ func (s *visibilityStore) serializePageToken(token *visibilityPageToken) ([]byte
 	return data, nil
 }
 
-func (s *visibilityStore) convertSearchResultToVisibilityRecord(hit *elastic.SearchHit) *persistence.VisibilityWorkflowExecutionInfo {
-	var source *visibilityRecord
-	// TODO (alex): use hit.Fields instead
-	err := json.Unmarshal(hit.Source, &source)
-	if err != nil { // log and skip error
-		s.logger.Error("unable to unmarshal search hit source",
-			tag.Error(err), tag.ESDocID(hit.Id))
+func (s *visibilityStore) generateESDoc(request *persistence.InternalVisibilityRequestBase, visibilityTaskKey string) map[string]interface{} {
+	doc := map[string]interface{}{
+		searchattribute.VisibilityTaskKey: visibilityTaskKey,
+		searchattribute.NamespaceID:       request.NamespaceID,
+		searchattribute.WorkflowID:        request.WorkflowID,
+		searchattribute.RunID:             request.RunID,
+		searchattribute.WorkflowType:      request.WorkflowTypeName,
+		searchattribute.StartTime:         request.StartTimestamp.UnixNano(),
+		searchattribute.ExecutionTime:     request.ExecutionTimestamp.UnixNano(),
+		searchattribute.ExecutionStatus:   request.Status,
+		searchattribute.TaskQueue:         request.TaskQueue,
+	}
+
+	if len(request.Memo.GetData()) > 0 {
+		doc[searchattribute.Memo] = request.Memo.GetData()
+		doc[searchattribute.Encoding] = request.Memo.GetEncodingType().String()
+	}
+
+	typeMap, err := s.searchAttributesProvider.GetSearchAttributes(s.index, false)
+	if err != nil {
+		s.logger.Error("Unable to read search attribute types.", tag.Error(err))
+	}
+
+	searchAttributes, err := searchattribute.Decode(request.SearchAttributes, &typeMap)
+	if err != nil {
+		s.logger.Error("Unable to decode search attributes.", tag.Error(err))
+		s.metricsClient.IncCounter(metrics.ElasticSearchVisibility, metrics.ESInvalidSearchAttribute)
+	}
+	for saName, saValue := range searchAttributes {
+		doc[saName] = saValue
+	}
+
+	return doc
+}
+
+func (s *visibilityStore) parseESDoc(hit *elastic.SearchHit, saTypeMap searchattribute.NameTypeMap) *persistence.VisibilityWorkflowExecutionInfo {
+	logUnexpectedType := func(fieldName string, fieldValue interface{}) {
+		s.logger.Error("Unexpected field type while parsing Elasticsearch document.", tag.Name(fieldName), tag.ValueType(fieldValue), tag.ESDocID(hit.Id))
+		s.metricsClient.IncCounter(metrics.ElasticSearchVisibility, metrics.ESInvalidSearchAttribute)
+	}
+	logNumberParseError := func(fieldName string, fieldValue json.Number, err error) {
+		s.logger.Error("Unable to parse JSON number while parsing Elasticsearch document.", tag.Name(fieldName), tag.ValueType(fieldValue), tag.Error(err), tag.ESDocID(hit.Id))
+		s.metricsClient.IncCounter(metrics.ElasticSearchVisibility, metrics.ESInvalidSearchAttribute)
+	}
```

Comment:
move these out as dedicated function? if these 2 fn are not capturing anything?

Reply:
e.g. making `hit.Id` part of the function input?

Reply:
It is more about encapsulation. I don't want to expose func to entire package if it is used only in one other func. Are you against this pattern?

Reply:
my guess, capture can create too many GC? plus these functions are internal to the package, which is controlled by you

Reply:
I always try to avoid captures, it is a good idea to replace it with param here. I would like to keep functions internally though.